.github/workflows/deliver.yml

```diff
@@ -155,3 +155,54 @@ jobs:
           files: /tmp/farmOS-${{ env.FARMOS_VERSION }}.tar.gz
           draft: false
           prerelease: false
+  publish:
+    name: Publish to Docker Hub
+    # We only publish to Docker Hub if this is a tag or 2.x branch push event
+    # to the official repository.
+    if: github.repository == 'farmOS/farmOS' && github.event_name == 'push' && (github.ref_type == 'tag' || (github.ref_type == 'branch' && needs.build.outputs.farmos_version == '2.x'))
+    runs-on: ubuntu-latest
+    needs:
+      - build
+      - sniff
+      - test
+    steps:
+      - name: Set FARMOS_VERSION from previous output.
+        run: echo "FARMOS_VERSION=${{ needs.build.outputs.farmos_version }}" >> $GITHUB_ENV
+      - name: Restore farmOS 2.x Docker image from cache
+        uses: actions/cache@v3
+        with:
+          path: /tmp/farmos-2x.tar
+          key: farmos-2x-${{ github.run_id }}
+      - name: Load farmos/farmos:2.x image
+        run: docker load < /tmp/farmos-2x.tar
+      - name: Restore farmOS 2.x-dev Docker image from cache
+        uses: actions/cache@v3
+        with:
+          path: /tmp/farmos-2x-dev.tar
+          key: farmos-2x-dev-${{ github.run_id }}
+      - name: Load farmos/farmos:2.x-dev image
+        run: docker load < /tmp/farmos-2x-dev.tar
+      - name: Login to Docker Hub.
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+      # If the 2.x branch was pushed...
+      - name: Publish farmos/farmos:2.x image to Docker Hub.
+        if: github.ref_type == 'branch' && env.FARMOS_VERSION == '2.x'
+        run: docker push farmos/farmos:2.x
+      - name: Publish farmos/farmos:2.x-dev image to Docker Hub.
+        if: github.ref_type == 'branch' && env.FARMOS_VERSION == '2.x'
+        run: docker push farmos/farmos:2.x-dev
```

Comment:
It looks like the `2.x` and `2.x-dev` docker tags will always have the same image - as opposed to the `2.x` tag only being updated for tagged releases. Am I reading that right? If so, does that match user expectations?

Reply:
`docker push farmos/farmos:2.x` will push the `2.x` image and `docker push farmos/farmos:2.x-dev` will push the `2.x-dev` image. Both get pushed whenever the `2.x` branch gets new commits. So I think this is correct, unless I'm overlooking something?

Reply:
Isn't the `2.x` docker tag a little redundant - either with the `2.x-dev` or `latest` docker tags? It seems like it could be a source of issues for folks who unintentionally end up running pre-release versions when they mean to be using `latest` (and actually probably should be using tagged releases)?

I don't think it needs to block this PR or anything since I suspect this is in line with the existing behavior from Docker Hub builds, I'm just asking the question whether it actually makes sense to keep the `2.x` tag when there's also the `2.x-dev` and `latest` tags that are more explicit.

Reply:
Oh gotcha - yea this just maintains the existing behavior, which is also a continuation of the old 1.x behavior, which was defined in the build rules on our Docker Hub.

The `2.x` image is different from the `2.x-dev` image, which includes things like XDebug. It is the difference between these two `Dockerfile`s:

- https://github.com/farmOS/farmOS/blob/2.x/docker/Dockerfile
- https://github.com/farmOS/farmOS/blob/2.x/docker/dev/Dockerfile

The `latest` image will always be the same as the most recent semantic version tag.

So the `2.x` image is distinct from both `2.x-dev` and `latest`, in that it is the _non-dev_ image with the latest commit of the `2.x` branch. Someone _may_ want to use this if they need a commit that hasn't been included in a tagged release yet, but don't want dev dependencies included. Although hopefully in practice this won't be common.

Reply:
Oh right! Forgot about the dev vs non-dev distinction. That makes sense!